README.md
# RaspberryPi-ArduinoBattery
## Add-on: Display Battery indicator on libreelec(kodi)

### Installation
  1. Download `service.arduino-battery.zip` to your libreELEC shared folder
  2. Install as addon (install from .zip file) and choose downloaded `service.arduino-battery.zip`
  3. Enjoy!

![Screenshot one](https://github.com/djspawnbrest/RaspberryPi-ArduinoBattery/blob/libreelec/screenshot1.png)
##
![Screenshot two](https://github.com/djspawnbrest/RaspberryPi-ArduinoBattery/blob/libreelec/screenshot2.png)


service.arduino-battery/changelog.txt
1.0.1 The first release shows the battery icon in the bottom-right corner.
1.0.0 Initial Release

src/ardbatt.c
#include <lxpanel/plugin.h>

#include <linux/i2c-dev.h>
#include <gtk/gtk.h>
#include <fcntl.h>
#include <sys/ioctl.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <time.h>
#include <glib.h>
#include <glib/gi18n.h>
#include <string.h>

typedef struct 
{
  char *delay_in_ms_str;
  int delay_in_ms;
  unsigned int timer;
  GtkWidget *prg;
  GtkWidget *vbox;
  GtkWidget *main;
  GtkWidget *label;
} ArdBattApp;

  // Arduino Vref (measure)
  const float VRef = 1.082;
  // IN <-[ Rh ]--(analogInPin)--[ Rl ]--|GND
  const float Rh = 46.87; // kOm
  const float Rl = 10.11; // kOm
  const float cell_min = 3.20; // min charge cell
  const float cell_max = 4.10; // max charge cell
  int AnalogVal;
  float VoltMult;
  float Volt;
  int Percent;
  int fd;
	char *fileName = "/dev/i2c-1";
	int  address = 0x20;
  int flag20 = 1;
  int status;

static int update_display(ArdBattApp *pArdBatt) {
  
  unsigned char buf[3];
	read(fd, buf, 3);
	VoltMult = (Rh + Rl) / Rl;
	AnalogVal = buf[0] << 8 | buf[1];
	Volt = AnalogVal * VoltMult * VRef / 1023;
	Percent = ((Volt - cell_min) * 100) / (cell_max - cell_min);
	if (Percent < 0) {
		Percent = 0;
	} else if (Percent > 100) {
		Percent = 100;
	}
  
  char *chg = "";
  
  if ( buf[2] ) {
    chg = "CHG";
  }
  
  double fraction = Percent/100.00;
  char prgBuf[5] = {'\0'};
  snprintf(prgBuf, sizeof(prgBuf), "%d%%", Percent);
  gtk_progress_bar_set_text(GTK_PROGRESS_BAR(pArdBatt->prg), prgBuf);
  gtk_progress_bar_set_fraction (GTK_PROGRESS_BAR(pArdBatt->prg), fraction);

  // make a label out of the ID
  char labBuf[25] = {'\0'};
  snprintf(labBuf, sizeof(labBuf), "%.2fV %s", Volt, chg);
  gtk_label_set_markup(GTK_LABEL(pArdBatt->label), labBuf);
  
  if (Percent < 20){
    if (flag20 == 1) {
      flag20 = 0;
      system("zenity --warning --title='Battery low level WARNING' --width=100 --height=100 --no-wrap --text='<big><b>Battery level is too low:</b></big>\n\nBattery level less than 20%!\nPlease <b>CONNECT</b> the charger or your system will be shuted down soon!'");
    }
  }
  
  if (Percent > 28) {
    flag20 = 1;
  }
  
  return 1;
}

static void set_delay_in_ms_str(ArdBattApp *pArdBatt) {
  free(pArdBatt->delay_in_ms_str);
  pArdBatt->delay_in_ms_str = malloc(8);
  snprintf(pArdBatt->delay_in_ms_str, 8, "%d", pArdBatt->delay_in_ms);
}

static void update_timer(ArdBattApp *pArdBatt) {
  if (pArdBatt->delay_in_ms_str != NULL)
    pArdBatt->delay_in_ms = atoi(pArdBatt->delay_in_ms_str);
  if (pArdBatt->timer != 0) {
    g_source_remove(pArdBatt->timer);
  }
  if (pArdBatt->delay_in_ms <= 10) {
    pArdBatt->delay_in_ms = 1000;
    set_delay_in_ms_str(pArdBatt);
  }
  if (pArdBatt->delay_in_ms > 2000 || pArdBatt->delay_in_ms % 1000 == 0)
    pArdBatt->timer = g_timeout_add_seconds(pArdBatt->delay_in_ms / 1000, 
                                      (GSourceFunc) update_display,
                                      (gpointer) pArdBatt);
  else
    pArdBatt->timer= g_timeout_add(pArdBatt->delay_in_ms, 
                    (GSourceFunc) update_display, (gpointer) pArdBatt);
}

GtkWidget *test_constructor(LXPanel *panel, config_setting_t *settings)
{
  (void)panel;
  (void)settings;
 
  ArdBattApp *pArdBatt = g_new0(ArdBattApp, 1);
 
  if (pArdBatt->delay_in_ms == 0)
    pArdBatt->delay_in_ms = 1000;
  set_delay_in_ms_str(pArdBatt);
  
  fd = open(fileName, O_RDWR);
  ioctl(fd, I2C_SLAVE, address);
  
  //---------------------- STYLES ----------------------------
  PangoFontDescription *prbr;
  prbr = pango_font_description_new();
  pango_font_description_set_family(prbr, "Arial");
  pango_font_description_set_size(prbr, 7 * PANGO_SCALE);
  PangoAttrList *attrlist;
  PangoAttribute *attr;
  PangoFontDescription *df;
  attrlist = pango_attr_list_new();
  df = pango_font_description_new();
  pango_font_description_set_family(df, "Arial");
  pango_font_description_set_size(df, 5/*<-fontsize*/ * PANGO_SCALE);
  attr = pango_attr_font_desc_new(df);
  pango_font_description_free(df);
  pango_attr_list_insert(attrlist, attr);
  gtk_label_set_attributes(GTK_LABEL(pArdBatt->label), attrlist);
  gtk_label_set_justify(GTK_LABEL(pArdBatt->label), GTK_JUSTIFY_CENTER);
  gtk_widget_set_size_request(GTK_WIDGET(pArdBatt->label), -1, -1);
  GtkRequisition size;
  gtk_widget_set_size_request(GTK_WIDGET(pArdBatt->label), size.width, -1);
  //---------------------------------------------------------

  pArdBatt->label = gtk_label_new("");
  gtk_widget_show(pArdBatt->label);
 
  pArdBatt->vbox = gtk_vbox_new (FALSE, 2);
  gtk_container_set_border_width (GTK_CONTAINER(pArdBatt->vbox), 2);
  gtk_widget_set_size_request(pArdBatt->vbox, -1, 25);
  gtk_widget_show(pArdBatt->vbox);
 
  pArdBatt->prg = gtk_progress_bar_new();
  gtk_widget_modify_font(pArdBatt->prg, prbr);
  gtk_widget_set_size_request(pArdBatt->prg, -1, 12);
  gtk_progress_bar_pulse (GTK_PROGRESS_BAR (pArdBatt->prg));
  gtk_widget_show(pArdBatt->prg);

  pArdBatt->main = gtk_event_box_new();

  gtk_widget_set_has_window(pArdBatt->main, FALSE);

  lxpanel_plugin_set_data(pArdBatt->main, pArdBatt, g_free);

  gtk_container_set_border_width(GTK_CONTAINER(pArdBatt->main), 2);

  gtk_container_add(GTK_CONTAINER (pArdBatt->main), pArdBatt->vbox);
  gtk_box_pack_start(GTK_BOX(pArdBatt->vbox), pArdBatt->prg, TRUE, FALSE, 0);
  gtk_box_pack_end(GTK_BOX(pArdBatt->vbox), pArdBatt->label, TRUE, TRUE, 0);

  gtk_widget_set_size_request(pArdBatt->main, 80, -1);

  // updates
  update_display(pArdBatt);

  update_timer(pArdBatt);
 
  return pArdBatt->main;
}

FM_DEFINE_MODULE(lxpanel_gtk, test)

/* Plugin descriptor. */
LXPanelPluginInit fm_module_init_lxpanel_gtk = {
  .name = "Arduino Battery by Spawn",
  .description = "Arduino Battery indicator by Spawn",
  .new_instance = test_constructor,
};


service.arduino-battery/addon.py
import time
import os
import signal
import subprocess
import xbmcaddon
import xbmcgui
import sys
sys.path.append('/storage/.kodi/addons/virtual.system-tools/lib')
import smbus

# Config
bus = smbus.SMBus(1) # init i2c bus
address = 0x20 # i2c arduino addres
iconState = ""
ADDON = xbmcaddon.Addon()
CWD = ADDON.getAddonInfo('path').decode("utf-8")
PNGVIEWPATH = xbmc.translatePath( os.path.join( CWD, 'resources', 'png' ).encode("utf-8") ).decode("utf-8")
VIDEOPATH = xbmc.translatePath( os.path.join( CWD, 'resources', 'video' ).encode("utf-8") ).decode("utf-8")
ICONPATH = xbmc.translatePath( os.path.join( CWD, 'resources', 'icons' ).encode("utf-8") ).decode("utf-8")

# Capacity config
# VBat <--[Rh]--(AnalogPin)--[Rl]--| GND
VRef = 1.082 # Vref arduino (must be measured)
Rh = 46.87 # kOm
Rl = 10.11 # kOm
cell_min = 3.20 # min charge cell (volts)
cell_max = 4.15 # max charge cell (volts)
oldPrc = None
oldChg = None
# Process kill flag
kill = False
# icon offset (px)
offset = 42

# chmod for pngview
os.system("chmod 755 " + PNGVIEWPATH + "/pngview")

def changeicon(percent, flag):
    global iconState
    global kill
    global offset
    if iconState != percent:
        iconsState = percent
        kill = True
        # Calculate screen width
        x = xbmcgui.getScreenWidth() - offset
        y = xbmcgui.getScreenHeight() - offset
        os.system(PNGVIEWPATH + "/pngview -b 0 -l 3000 -x " + str(x) + " -y " + str(y) + " " + ICONPATH + "/batt" + percent + "_" + flag + ".png &")

def changePerc(prc):
    result = 0
    if 95 <= prc <= 100:
        result = 100
    elif 85 <= prc <= 94:
        result = 90
    elif 75 <= prc <= 84:
        result = 80
    elif 65 <= prc <= 74:
        result = 70
    elif 55 <= prc <= 64:
        result = 60
    elif 45 <= prc <= 54:
        result = 50
    elif 35 <= prc <= 44:
        result = 40
    elif 25 <= prc <= 34:
        result = 30
    elif 15 <= prc <= 24:
        result = 20
    elif 5 <= prc <= 14:
        result = 10
    elif 0 <= prc <= 4:
        result = 0
    return result

if __name__ == '__main__':
    while not xbmc.abortRequested:
        # code
        try:
            VoltMult = (Rh + Rl) / Rl
            buf = bus.read_i2c_block_data(address,0x00,3) # read values from i2c (arduino)
            AnalogVal = buf[0] << 8 | buf[1] # convert byte values to Vbat analog value from arduino
            Volt = AnalogVal * VoltMult * VRef / 1023 # convert in battery voltage
            prcnt = int(round( ((Volt - cell_min) * 100) / (cell_max - cell_min) )) # convert battery voltage to Capacity
            if prcnt < 0:
                prcnt = 0
            elif prcnt > 100:
                prcnt = 100

            res = changePerc(prcnt) # get icon Capacity value

            if res != oldPrc or buf[2] != oldChg:
                oldChg = buf[2]
                oldPrc = res
                if kill:
                    kill = False
                    os.system("killall pngview")

                chg = str(int(buf[2])) # read charger flag (charger - 1 | 0 - not)
                changeicon(str(res), chg)

            if res == 0 and int(buf[2]) == 0:
                xbmc.executebuiltin('XBMC.PlayMedia(' + VIDEOPATH + '/timer.mp4)')
                time.sleep(60)
                os.system("shutdown -h now")

        except:
            pass
            time.sleep(5) # sleep 5 seconds
        xbmc.sleep(2000) # 1000 - 1 time on second

main.py
import os
import sys
import time
import smbus
import signal
import subprocess
from subprocess import check_output
from config import *

# Global vars
bus = smbus.SMBus(1)
WARNING = 0
LAYER = 299999
DICTIONNARY = dict({"batt":0, })

# Get Pathes
PATH = sys.path[0]
PNGVIEWPATH = PATH + PNGVIEW
VIDEOPATH = PATH + VIDEOS
ICONPATH = PATH + ICONS

# Get the screen resolution on X and Y axis
XRESOLUTION = int(check_output("fbset | grep 'x' | tr 'a-z\"-' ' ' | awk '{printf $1;}'", shell=True))
YRESOLUTION = int(check_output("fbset | grep 'x' | tr 'a-z\"-' ' ' | awk '{printf $2;}'", shell=True))

# chmod for pngview
os.system("chmod 755 " + PNGVIEWPATH + "/pngview")

# Debug
if DEBUGMSG:
    print("Screen resolution :          {}x{}".format(XRESOLUTION,YRESOLUTION))

# Positions for the choosen corner
if CORNER == 1:
	XPOS = XOFFSET
	YPOS = YOFFSET
elif CORNER == 2:
	XPOS = XRESOLUTION - XOFFSET
	YPOS = YOFFSET
elif CORNER == 3:
	XPOS = XRESOLUTION - XOFFSET
	YPOS = YRESOLUTION - YOFFSET
elif CORNER == 4:
	XPOS = XOFFSET
	YPOS = YRESOLUTION - YOFFSET

def changeIcon(icon, state, chg):
	global LAYER
	global DICTIONNARY

	#LAYER -= 1 # Useless ???
	cmd = '{}/pngview'.format(PNGVIEWPATH)
	arg1 = '-b 0'
	arg2 = '-l {}'.format(LAYER)
	arg3 = '-x {}'.format(XPOS)
	arg4 = '-y {}'.format(YPOS)
	img = '{}/{}{}_{}.png'.format(ICONPATH, icon, state, chg)

	# New image
	popen = subprocess.Popen([cmd, arg1, arg2, arg3, arg4, img])
	time.sleep(2)

	# Get the last id of pngview for this specific icon
	pid = DICTIONNARY[icon]
	# Remove previous icon if it exist
	if pid != 0:
		os.kill(pid, signal.SIGTERM)
	# Get the process id of the new image
	new_pid = popen.pid
	# Update the dictionnary
	if new_pid != "":
		DICTIONNARY[icon] = new_pid
		
	# Debug
	if DEBUGMSG:
		print("Pngview previous pid was {} for {} icon".format(pid, icon))
		print("Change {} icon to {}".format(icon, state))
		print("Pngview pid is now {} for {} icon".format(new_pid, icon))

def playClip(clip):
	global WARNING
	global VIDEOPATH
	if clip == "alert" and WARNING != 1:
		os.system("/usr/bin/omxplayer --no-osd --layer 999999 " + VIDEOPATH + "/alert.mp4 --alpha 160")
		WARNING = 1
	elif clip == "shutdown":
		os.system("/usr/bin/omxplayer --no-osd --layer 999999 " + VIDEOPATH + "/timer.mp4 --alpha 160;shutdown -h now")

def changePerc(prc):
    result = 0
    if 95 <= prc <= 100:
        result = 100
    elif 85 <= prc <= 94:
        result = 90
    elif 75 <= prc <= 84:
        result = 80
    elif 65 <= prc <= 74:
        result = 70
    elif 55 <= prc <= 64:
        result = 60
    elif 45 <= prc <= 54:
        result = 50
    elif 35 <= prc <= 44:
        result = 40
    elif 25 <= prc <= 34:
        result = 30
    elif 15 <= prc <= 24:
        result = 20
    elif 5 <= prc <= 14:
        result = 10
    elif 0 <= prc <= 4:
        result = 0
    return result

def checkBattery():
	try:
		global WARNING
		VoltMult = (Rh + Rl) / Rl
		buf = bus.read_i2c_block_data(address,0x00,3)
		AnalogPin = buf[0] << 8 | buf[1]
		voltage = AnalogPin * VoltMult * VRef / 1023
		prcnt = int(round( ((voltage - cell_min) * 100) / (cell_max - cell_min) ))
		if prcnt < 0:
			prcnt = 0
		elif prcnt > 100:
			prcnt = 100
		res = changePerc(prcnt)

		if DEBUGMSG:
			print("ADC value: {} ({}V)".format(AnalogPin, voltage))

		# Battery monitor
		chg = str(int(buf[2])) # read charger flag (charger - 1 | 0 - not)
		changeIcon("batt", res, chg)

		if prcnt > 25:
			WARNING = 0

		if 10 <= prcnt <= 20:
			playClip("alert")
		elif prcnt <= 9 and int(buf[2]) != 1:
			playClip("shutdown")

	except:
		pass
		if DEBUGMSG:
			print("ERR")
		time.sleep(2)

def endProcess(signalnum = None, handler = None):
    os.system("killall pngview")
    exit(0)

# Prepare handlers for process exit
signal.signal(signal.SIGTERM, endProcess)
signal.signal(signal.SIGINT, endProcess)

# Main function
while True:
    checkBattery()

    # Debug
    if DEBUGMSG:
        print("GLOBAL Variables are : \nWARNING = {}\nLAYER = {}\n PID = {}".format(WARNING, LAYER, DICTIONNARY))

    # Wainting for next loop
    time.sleep(REFRESH_RATE)
